### PR TITLE
[Refactor] GLM-ASR Modeling

### PR DESCRIPTION
**Pull request by JaredforReal**
### 🚀 Key Improvements

**1. Native vLLM Audio Encoder Implementation** (`glmasr.py`)
- Completely rewrote `GlmAsrEncoder` as a vLLM-native implementation with full optimization support:
  - **QKVParallelLinear**: Fused Q/K/V projections for efficient attention computation
  - **ColumnParallelLinear/RowParallelLinear**: Tensor parallelism support for distributed inference
  - **Quantization support**: Compatible with vLLM's quantization framework
  - **Flash Attention (SDPA)**: Leverages PyTorch's `scaled_dot_product_attention` for optimized attention
- Implemented `GlmAsrRotaryEmbedding` with pre-computed cos/sin cache for faster RoPE computation
- Built `GlmAsrAttention`, `GlmAsrMLP`, and `GlmAsrEncoderLayer` with optimized layer norms and residual connections
- Proper grouped query attention (GQA) handling with k/v repetition

**2. Cleaner Architecture: Direct BaseMultiModalProcessor Inheritance** (`glmasr.py`)
- Refactored `GlmAsrMultiModalProcessor` to inherit directly from `BaseMultiModalProcessor`
- Refactored `GlmAsrProcessingInfo` to inherit directly from `BaseProcessingInfo`
- Refactored `GlmAsrMultiModalDataParse` to inherit directly from `BaseMultiModalDataParse`
- Removed dependency on `AudioFlamingo3` for cleaner, more maintainable code
- Streamlined processing pipeline with better performance and reduced complexity
- Maintained full compatibility with existing API and functionality

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/glmasr_utils.py [hunk 1]
```diff
@@ -1,19 +1,90 @@
...
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
+import logging
 from collections.abc import Sequence
 from typing import cast
 
 import torch
 import torch.nn as nn
 
+logger = logging.getLogger(__name__)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr_utils.py]: The logger seems unused

File: vllm/model_executor/models/glmasr.py [hunk 2]
```diff
@@ -73,6 +72,505 @@
 from .utils import AutoWeightsLoader, init_vllm_registered_model, maybe_prefix
 from .whisper import ISO639_1_SUPPORTED_LANGS
 
+logger = logging.getLogger(__name__)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: Same

File: vllm/model_executor/models/glmasr.py [hunk 3]
```diff
@@ -73,6 +72,505 @@
...
+        emb = torch.cat((freqs, freqs), dim=-1)
+
+        # Compute and cache cos/sin
+        cos = emb.cos() * self.attention_scaling
+        sin = emb.sin() * self.attention_scaling
+
+        self.register_buffer("cos_cached", cos, persistent=False)
+        self.register_buffer("sin_cached", sin, persistent=False)
+
+    def forward(
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: We should use `ApplyRotaryEmb` to apply rotary embeddings now. See #29873 on how to do it

File: vllm/model_executor/models/glmasr.py [hunk 4]
```diff
@@ -203,7 +724,12 @@ def _parse_audio_data(
         return super()._parse_audio_data(data)
 
 
-class GlmAsrMultiModalProcessor(AudioFlamingo3MultiModalProcessor):
+class GlmAsrMultiModalProcessor(BaseMultiModalProcessor["GlmAsrProcessingInfo"]):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py:729]: Let's not include any GPU logic in the MM processor for now.

Comment (DarkLight1337) [reply vllm/model_executor/models/glmasr.py:729]: Any GPU logic should be done based on `device` argument so it's easy to configure later. Also see #22070

Comment (DarkLight1337) [reply vllm/model_executor/models/glmasr.py:729]: Edit: I see that you only kept the CPU logic. But honestly I think this chunking logic should be part of HF processor as well

File: vllm/model_executor/models/glmasr_utils.py (hunk 1 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr_utils.py]: done

File: vllm/model_executor/models/glmasr.py (hunk 2 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: done

File: vllm/model_executor/models/glmasr.py (hunk 4 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py:729]: Comment deleted， working on it

**Comment by JaredforReal** [comment]
@DarkLight1337 PTAL, thanks!

File: vllm/model_executor/models/glmasr.py [hunk 5]
```diff
@@ -74,6 +71,497 @@
...
+                )
+                / dim
+            )
+        )
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+        # Pre-compute cos/sin cache for efficiency
+        self._set_cos_sin_cache(self.max_seq_len_cached, device)
+
+        # Use vLLM's ApplyRotaryEmb CustomOp for efficient rotary embedding
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: Seems unused

File: vllm/model_executor/models/glmasr.py [hunk 6]
```diff
@@ -74,6 +71,470 @@
 from .whisper import ISO639_1_SUPPORTED_LANGS
 
 
+# Optimized vLLM Native GlmAsrEncoder Implementation
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: I think this comment is a bit redundant tbh

File: vllm/model_executor/models/glmasr.py [hunk 7]
```diff
@@ -74,6 +71,470 @@
...
+        )
+
+        # Use vLLM's ApplyRotaryEmb CustomOp
+        # enforce_enable=True ensures the op is always enabled (important for ViT)
+        self.apply_rotary_emb = ApplyRotaryEmb(enforce_enable=True)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cos: torch.Tensor,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: Rename to `rotary_pos_emb_cos` etc. to be in line with the other models

File: vllm/model_executor/models/glmasr.py (hunk 6 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: removed

File: vllm/model_executor/models/glmasr.py (hunk 7 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: done, thanks！

**Comment by JaredforReal** [comment]
image

@DarkLight1337 tested, my benchmark right here:
image

File: vllm/model_executor/models/glmasr.py [hunk 8]
```diff
@@ -74,6 +71,471 @@
...
+            q,
+            k,
+            v,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=False,
+        )
+
+        # Reshape back
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.view(batch_size, seq_len, -1)
```
Comment (Isotr0py) [inline vllm/model_executor/models/glmasr.py]: I think we can use `MMEncoderAttention` here:
https://github.com/vllm-project/vllm/blob/2c1a4f2488da2bb3c60dcbbb5df61a370e39ee27/vllm/attention/layers/mm_encoder_attention.py#L22-L23

Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: done, and everything works fine.
image
image

File: vllm/model_executor/models/glmasr.py [hunk 9]
```diff
@@ -444,16 +1059,21 @@ def _process_audio_input(
...
-        return _group_audio_embeddings(chunk_embeddings, chunk_counts)
+        result = _group_audio_embeddings(chunk_embeddings, chunk_counts)
+
+        return result
 
     def get_language_model(self) -> torch.nn.Module:
         return self.language_model
 
     def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
+        ""
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]:
```suggestion
```

File: vllm/model_executor/models/glmasr.py [hunk 10]
```diff
@@ -444,16 +1059,21 @@ def _process_audio_input(
         chunk_embeddings = torch.split(
             masked_audio_features, audio_output_lengths.flatten().tolist()
         )
-        return _group_audio_embeddings(chunk_embeddings, chunk_counts)
+        result = _group_audio_embeddings(chunk_embeddings, chunk_counts)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: Unneccessary change

File: vllm/model_executor/models/glmasr.py (hunk 9 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: done

File: vllm/model_executor/models/glmasr.py (hunk 10 above)
Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: done

File: vllm/model_executor/models/glmasr.py [hunk 11]
```diff
@@ -295,22 +843,65 @@ def _get_prompt_updates(
...
+                counts_list = _as_list_chunk_counts(chunk_counts)
+                start_idx = 0
+                for count in counts_list:
+                    end_idx = start_idx + count
+                    if isinstance(feature_attention_mask, torch.Tensor):
+                        mask = feature_attention_mask[start_idx:end_idx]
+                    else:
+                        mask = feature_attention_mask[start_idx:end_idx]
+                        if isinstance(mask, list):
+                            mask = torch.stack(mask)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]:
```suggestion
                start_idx = 0
                for count in _as_list_chunk_counts(chunk_counts):
                    end_idx = start_idx + count

                    mask = feature_attention_mask[start_idx:end_idx]
                    if isinstance(mask, list):
                        mask = torch.stack(mask)
```

File: vllm/model_executor/models/glmasr.py [hunk 12]
```diff
@@ -295,22 +843,65 @@ def _get_prompt_updates(
...
+                    start_idx = end_idx
+            else:
+                # Single chunk per audio
+                for idx in range(len(feature_attention_mask)):
+                    if isinstance(feature_attention_mask, torch.Tensor):
+                        mask = feature_attention_mask[idx : idx + 1]
+                    else:
+                        mask = feature_attention_mask[idx]
+                        if not isinstance(mask, torch.Tensor):
+                            mask = torch.tensor(mask)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]:
```suggestion
                    mask = feature_attention_mask[idx : idx + 1]
                    if isinstance(mask, list):
                        mask = torch.stack(mask)
```

**Review by DarkLight1337 — APPROVED** [review]
LGTM now. @Isotr0py can you check as well?

**Review by Isotr0py — APPROVED** [review]
LGTM now

File: vllm/model_executor/models/glmasr.py [hunk 13]
```diff
@@ -775,11 +775,18 @@ def _call_hf_processor(
...
+
+        # Inject device if not specified - use current platform's device
+        from vllm.platforms import current_platform
+
+        device = mm_kwargs.get("device", current_platform.device_type)
+
         mm_kwargs = dict(
             **mm_kwargs,
             sampling_rate=feature_extractor.sampling_rate,
+            device=device,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: Is this really necessary? Since you can just pass `device` to `mm_kwargs`

Comment (DarkLight1337) [reply vllm/model_executor/models/glmasr.py]: I think we should avoid defaulting to using GPU acceleration

Comment (JaredforReal) [reply vllm/model_executor/models/glmasr.py]: in my experiment, building audio_kwargs is necessary to ensure device passed to transformers, but it seems like it affected a lot of accuracy……orz

**Comment by JaredforReal** [comment]
@DarkLight1337 @Isotr0py Back to the version you guys approved. Let's land it for now, Thanks!
